packages/mui-joy/src/IconButton/IconButton.tsx

```diff
@@ -0,0 +1,197 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { unstable_capitalize as capitalize, unstable_useForkRef as useForkRef } from '@mui/utils';
+import { useButton } from '@mui/base/ButtonUnstyled';
+import composeClasses from '@mui/base/composeClasses';
+import { styled, useThemeProps } from '../styles';
+import iconButtonClasses, { getIconButtonUtilityClass } from './iconButtonClasses';
+import { IconButtonProps, IconButtonTypeMap, ExtendIconButton } from './IconButtonProps';
+
+const useUtilityClasses = (ownerState: IconButtonProps & { focusVisible: boolean }) => {
+  const { color, disabled, focusVisible, focusVisibleClassName, size, shape, variant } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      variant && `variant${capitalize(variant)}`,
+      color && `color${capitalize(color)}`,
+      size && `size${capitalize(size)}`,
+      shape && `shape${capitalize(shape)}`,
+    ],
+  };
+
+  const composedClasses = composeClasses(slots, getIconButtonUtilityClass, {});
+
+  if (focusVisible && focusVisibleClassName) {
+    composedClasses.root += ` ${focusVisibleClassName}`;
+  }
+
+  return composedClasses;
+};
+
+const IconButtonRoot = styled('button', {
+  name: 'MuiIconButton',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: IconButtonProps }>(({ theme, ownerState }) => [
+  {
+    '--IconButton-size': '2.5rem',
+    padding: '0.25rem',
+    ...(ownerState.variant === 'outlined' && {
+      padding: 'calc(0.25rem - 1px)', // account for the border width
+    }),
+    ...(ownerState.size === 'sm' && {
+      '--IconButton-size': '2rem',
+      // same padding as size `md`
+    }),
+    ...(ownerState.size === 'lg' && {
+      '--IconButton-size': '3rem',
+      padding: '0.5rem',
+      ...(ownerState.variant === 'outlined' && {
+        padding: 'calc(0.5rem - 1px)', // account for the border width
```

Comment:
Where are we setting the border?

Reply:
I guess it comes from the `theme.variants`. In that case, shouldn't we read these value from the theme? What if it changes and it's not `1px` anymore.

Reply:
Good point. I have considered it but did not add until we have a use-case that the border width's will be larger than 1px. Will try to find a real example and do the change if I find any.

Reply:
> Where are we setting the border?

If we don't minus 1px the size of the outlined Button will be '34px' (because of border 1px top and bottom) while other variants are exactly '32px'.

Reply:
Adobe spectrum has 2px border width. https://react-spectrum.adobe.com/react-spectrum/Button.html

I will update the variant to make it dynamic via the variable.

Reply:
Why don't we use `box-sizing: border-box` and specify the height only?